File: backend/schemas/checkpoint.py
from typing import Optional, Dict, Any
from datetime import datetime
from .base import BaseSchema


class CheckpointSchema(BaseSchema):
    id: str
    project_id: str
    agent_id: str
    type: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    content_category: Optional[str] = None
    output: Optional[Dict[str, Any]] = None
    status: str
    feedback: Optional[str] = None
    resolved_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class CheckpointCreateSchema(BaseSchema):
    id: Optional[str] = None
    project_id: str
    agent_id: str
    type: Optional[str] = "review"
    title: Optional[str] = None
    description: Optional[str] = None
    content_category: Optional[str] = "document"
    output: Optional[Dict[str, Any]] = None
    status: Optional[str] = "pending"


class CheckpointResolveSchema(BaseSchema):
    resolution: str
    feedback: Optional[str] = None

File: langgraph-studio/src/types/checkpoint.ts
import type{components}from'./api-generated'

export type CheckpointSchema=components['schemas']['CheckpointSchema']
export type CheckpointResolveSchema=components['schemas']['CheckpointResolveSchema']

export type CheckpointStatus='pending'|'approved'|'rejected'|'revision_requested'

export type CheckpointType=
 |'concept_review'
 |'design_review'
 |'scenario_review'
 |'character_review'
 |'world_review'
 |'task_split_review'
 |'code_review'
 |'asset_review'
 |'integration_review'
 |'test_review'
 |'final_review'
 |'release_decision'

export interface Checkpoint{
 id:string
 projectId:string
 agentId:string
 type:string
 title:string
 description:string|null
 output:CheckpointOutput
 status:CheckpointStatus
 feedback:string|null
 resolvedAt:string|null
 createdAt:string
 updatedAt:string
}

export interface CheckpointOutput{
 type?:string
 format?:string
 documentType?:string
 summary?:string
 content?:string|Record<string,unknown>
 tokensUsed?:number
 generationTimeMs?:number
 previewUrl?:string
}

export interface CheckpointResolution{
 checkpointId:string
 resolution:'approved'|'rejected'|'revision_requested'
 feedback?:string
}

export interface CheckpointWithMeta extends Checkpoint{
 waitingTimeMinutes:number
 agentName:string
 phaseName:string
}
